### PR TITLE
Support "readiness groups" i.e. dependencies

### PR DESCRIPTION
**Pull request by jveski**
Implements the concept of readiness groups.

### CODE REVIEW

File: internal/controllers/reconciliation/controller.go [hunk 1]
```diff
@@ -105,9 +105,12 @@ func (c *Controller) Reconcile(ctx context.Context, req *reconstitution.Request)
 
 	var prev *reconstitution.Resource
 	if comp.Status.PreviousSynthesis != nil {
-		synRef.UUID = comp.Status.PreviousSynthesis.UUID
-		prev, _ = c.resourceClient.Get(ctx, synRef, &req.Resource)
+		prevSynRef := reconstitution.NewSynthesisRef(comp)
```
Comment (jveski) [inline internal/controllers/reconciliation/controller.go:108]: Avoid mutating the synthesis ref struct by just constructing a new one first here.

File: internal/reconstitution/queueprocessor.go [hunk 2]
```diff
@@ -30,7 +30,7 @@ func (q *queueProcessor) processQueueItem(ctx context.Context) bool {
 	}
 	defer q.Queue.Done(item)
 
-	req, ok := item.(*Request)
+	req, ok := item.(Request)
```
Comment (jveski) [inline internal/reconstitution/queueprocessor.go:33]: Kind of unrelated change: using pointers breaks the workqueues de-duplication since it relies on equality checks.

File: internal/reconstitution/controller.go [hunk 3]
```diff
@@ -105,3 +117,45 @@ func (r *controller) populateCache(ctx context.Context, comp *apiv1.Composition,
...
+			SliceName: slice.Name,
+			Namespace: slice.Namespace,
+		})
+		if !ok {
+			logger.V(1).Info("a dependent resource was not found in cache - this is unexpected")
+			return ctrl.Result{}, nil
+		}
+
+		synRef := &SynthesisRef{CompositionName: owner.Name, Namespace: req.Namespace, UUID: slice.Spec.SynthesisUUID}
+		resources := r.Cache.RangeByReadinessGroup(ctx, synRef, res.ReadinessGroup, 1)
```
Comment (AYM1607) [inline internal/reconstitution/controller.go:150]: Const (and maybe type?) for the direction? It's a bit hard to read at first glance

File: internal/controllers/reconciliation/controller.go [hunk 4]
```diff
@@ -146,17 +138,49 @@ func (c *Controller) Reconcile(ctx context.Context, req *reconstitution.Request)
...
+	if (status == nil || !status.Reconciled) && !resource.Deleted() {
+		dependencies := c.resourceClient.RangeByReadinessGroup(ctx, synRef, resource.ReadinessGroup, -1)
+		for _, dep := range dependencies {
+			slice := &apiv1.ResourceSlice{}
+			err = c.client.Get(ctx, dep.ManifestRef.Slice, slice)
+			if err != nil {
+				return ctrl.Result{}, fmt.Errorf("getting resource slice: %w", err)
+			}
+			status := dep.FindStatus(slice)
+			if status == nil || !status.Reconciled {
```
Comment (AYM1607) [inline internal/controllers/reconciliation/controller.go:161]: Reconciled -> Ready?

File: internal/controllers/reconciliation/controller.go [hunk 5]
```diff
@@ -146,17 +138,49 @@ func (c *Controller) Reconcile(ctx context.Context, req *reconstitution.Request)
...
 		if ok {
 			ready = &readiness.ReadyTime
 		}
+	} else {
+		ready = status.Ready
+	}
+
+	// Evaluate the readiness of resources in the previous readiness group
+	if (status == nil || !status.Reconciled) && !resource.Deleted() {
+		dependencies := c.resourceClient.RangeByReadinessGroup(ctx, synRef, resource.ReadinessGroup, -1)
```
Comment (AYM1607) [inline internal/controllers/reconciliation/controller.go:153]: Correct me if I'm wrong, but the way I'm reading this and the indexing PR, these are just dependencies within the same composition; will we also have readiness groups across compositions?

Comment (AYM1607) [reply internal/controllers/reconciliation/controller.go:153]: I totally see the use of intra-composition dependencies (get etcd ready before api-server in control-plane-core for ex) but we should also run through the possible scenarios where a composition might depend on another one

Comment (jveski) [reply internal/controllers/reconciliation/controller.go:153]: This is an interesting thought - let's discuss elsewhere

File: internal/controllers/reconciliation/controller.go (hunk 4 above)
Comment (jveski) [reply internal/controllers/reconciliation/controller.go]: Good catch!

**Review by AYM1607 — APPROVED** [review]
LGTM!

File: internal/controllers/reconciliation/controller.go (hunk 5 above)
Comment (AYM1607) [reply internal/controllers/reconciliation/controller.go]: Discussed offline, will stick with intra-composition readiness for now, we'll evaluate if inter-composition is required in the future
